spark/src/main/scala/org/apache/spark/sql/delta/files/TahoeFileIndex.scala

```diff
@@ -126,6 +127,19 @@ abstract class TahoeFileIndex(
     addFile.defaultRowCommitVersion.foreach(defaultRowCommitVersion =>
       metadata.put(DefaultRowCommitVersion.METADATA_STRUCT_FIELD_NAME, defaultRowCommitVersion))
 
+    // Attach filter types to FileActions, so that later [[DeltaParquetFileFormat]] could pick it up
+    // to decide which kind of rows should be filtered out. This info is necessary for reading CDC
+    // rows that have been deleted (marked in DV), in which case marked rows must be kept rather
+    // than filtered out. In such a case, the `filterTypes` map will be populated by [[CDCReader]]
+    // to indicate IF_NOT_CONTAINED filter should be used. In other cases, `filterTypes` will be
+    // empty, so we generate IF_CONTAINED as the default DV behavior
+    if (addFile.deletionVector != null) {
+      metadata.put(DeltaParquetFileFormat.FILE_ROW_INDEX_FILTER_ID_ENCODED,
+        JsonUtils.toJson(addFile.deletionVector))
+    }
+    val filterType = rowIndexFilters.getOrElse(Map.empty)
+      .getOrElse(addFile.path, RowIndexFilterType.IF_CONTAINED)
+    metadata.put(DeltaParquetFileFormat.FILE_ROW_INDEX_FILTER_TYPE, filterType)
```

Comment:
This will add one metadata entry regardless the existence of DVs on thi file, right?
How about not adding it but later we do `partitionedFile.otherConstantMetadataColumnValues.contains(FILE_ROW_INDEX_FILTER_TYPE)`?

Reply:
@xupefei True that

Reply:
Although the map.get and isDefined also do the `contains` trick